compilers/concrete-compiler/compiler/lib/Dialect/FHELinalg/IR/FHELinalgOps.cpp

```diff
@@ -1687,6 +1687,488 @@ void ToUnsignedOp::getCanonicalizationPatterns(
   getSignedConvCanonicalizationPatterns<ToUnsignedOp>(patterns, context);
 }
 
+std::optional<mlir::Value>
+fuseBackToBackTableLookups(mlir::Value value, mlir::PatternRewriter &rewriter) {
+
+  using mlir::concretelang::FHE::FheIntegerInterface;
+
+  auto currentOperation = value.getDefiningOp();
+
+  auto currentOperationIsTlu =
+      llvm::isa<ApplyLookupTableEintOp>(currentOperation);
+  auto currentOperationIsMappedTlu =
+      llvm::isa<ApplyMappedLookupTableEintOp>(currentOperation);
+
+  if (!currentOperationIsTlu && !currentOperationIsMappedTlu) {
+    return std::nullopt;
+  }
+
+  auto intermediateValue =
+      (currentOperationIsTlu
+           ? (llvm::dyn_cast<ApplyLookupTableEintOp>(currentOperation).getT())
+           : (llvm::dyn_cast<ApplyMappedLookupTableEintOp>(currentOperation)
+                  .getT()));
+
+  auto intermediateOperation = intermediateValue.getDefiningOp();
+  if (!intermediateOperation) {
+    return std::nullopt;
+  }
+
+  auto intermediateOperationIsTlu =
+      llvm::isa<ApplyLookupTableEintOp>(intermediateOperation);
+  auto intermediateOperationIsMappedTlu =
+      llvm::isa<ApplyMappedLookupTableEintOp>(intermediateOperation);
+
+  if (!intermediateOperationIsTlu && !intermediateOperationIsMappedTlu) {
```

Comment:
All this can be collapsed by doing `dyn_cast_or_null` calls directly on `intermediateValue.getDefiningOp()`.